plugins/badges-backend/src/service/router.ts

```diff
@@ -0,0 +1,125 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import express from 'express';
+import Router from 'express-promise-router';
+import {
+  errorHandler,
+  PluginEndpointDiscovery,
+} from '@backstage/backend-common';
+import { CatalogApi, CatalogClient } from '@backstage/catalog-client';
+import { Config } from '@backstage/config';
+import { NotFoundError } from '@backstage/errors';
+import { BadgeBuilder, DefaultBadgeBuilder } from '../lib/BadgeBuilder';
+import { BadgeContext, BadgeFactories } from '../types';
+
+export interface RouterOptions {
+  badgeBuilder?: BadgeBuilder;
+  badgeFactories?: BadgeFactories;
+  catalog?: CatalogApi;
+  config: Config;
+  discovery?: PluginEndpointDiscovery;
+}
+
+export async function createRouter(
+  options: RouterOptions,
+): Promise<express.Router> {
+  if (!options.catalog && !options.discovery) {
+    throw new Error('must provide either catalog api or discovery api');
+  }
+  const catalog =
+    options.catalog || new CatalogClient({ discoveryApi: options.discovery! });
+  const badgeBuilder =
+    options.badgeBuilder ||
+    new DefaultBadgeBuilder(options.badgeFactories || {});
+  const router = Router();
+
+  router.get('/entity/:namespace/:kind/:name/badge-specs', async (req, res) => {
+    const { namespace, kind, name } = req.params;
+    const entity = await catalog.getEntityByName({ namespace, kind, name });
+    if (!entity) {
+      throw new NotFoundError(
+        `No ${kind} entity in ${namespace} named "${name}"`,
+      );
+    }
+
+    const context: BadgeContext = {
+      badgeUrl: '',
+      config: options.config,
+      entity,
+    };
+
+    const specs = [];
+    for (const badgeId of await badgeBuilder.getBadgeIds()) {
+      context.badgeUrl = [
+        `${req.protocol}://`,
+        req.headers.host,
+        req.originalUrl.replace(/badge-specs$/, badgeId),
+      ].join('');
+      const badge = await badgeBuilder.createBadge({
+        badgeId,
+        context,
+        format: 'json',
+      });
+
+      if (badge) {
+        specs.push(badge);
+      }
+    }
+
+    res.setHeader('Content-Type', 'application/json');
+    res.status(200).send(`[${specs.join(',\n')}]`);
+  });
+
+  router.get('/entity/:namespace/:kind/:name/:badgeId', async (req, res) => {
+    const { namespace, kind, name, badgeId } = req.params;
+    const entity = await catalog.getEntityByName({ namespace, kind, name });
+    if (!entity) {
+      throw new NotFoundError(
+        `No ${kind} entity in ${namespace} named "${name}"`,
+      );
+    }
+
+    let format =
+      req.accepts(['image/svg+xml', 'application/json']) || 'image/svg+xml';
+    if (req.query.format === 'json') {
+      format = 'application/json';
+    }
+
+    const badgeUrl = [
+      `${req.protocol}://`,
+      req.headers.host,
+      req.originalUrl,
+    ].join('');
```

Comment:
same here, munging the request url feels shaky, just form a new url from the resolver and a hard coded known trailing path on that